Wide-layout attachments: Fix baseline computation

squelart opened the pull request:
#### 28021b9b53fe677af02c6099ef04bfc612d6f673
<pre>
Wide-layout attachments: Fix baseline computation
<a href="https://bugs.webkit.org/show_bug.cgi?id=259039">https://bugs.webkit.org/show_bug.cgi?id=259039</a>
rdar://problem/111991128

Reviewed by Tim Nguyen.

The previous computation used `renderRect`s, which relied on the baseline, modifying the baseline again
and again.
Instead, assume that the image is vertically centred, so it&apos;s just a matter of adding half-heights to find
the bottom of that image. (Fallback to the full attachment height alone in case the inner element doesn&apos;t
have its height yet.)

* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/rendering/RenderAttachment.cpp:
(WebCore::RenderAttachment::baselinePosition const):

Canonical link: <a href="https://commits.webkit.org/265894@main">https://commits.webkit.org/265894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab36b2f28f074884911cc9749745a446032f8020

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12185 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12530 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12832 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13930 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11758 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12215 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14946 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12545 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14437 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12349 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13152 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10319 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14346 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10437 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11060 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18173 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11518 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11220 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14402 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11713 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9660 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10930 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2989 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15256 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11567 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderAttachment.cpp:
```diff
@@ -97,16 +97,12 @@ void RenderAttachment::layout()
-            return baselineElementTop + baselineElementRenderBox->baselinePosition(fontBaseline, firstLine, lineDirectionMode, linePositionMode);
-        return baselineElementTop + baselineElementRect.height();
+        if (auto* baselineElementRenderBox = baselineElement->renderBox())
+            return (height() + baselineElementRenderBox->height()) / 2;
```
```suggestion
            // This is the bottom of the icon assuming it is vertically centered.
            return (height() + baselineElementRenderBox->height()) / 2;
```

nt1m commented on Source/WebCore/rendering/RenderAttachment.cpp:
```diff
@@ -97,16 +97,12 @@ void RenderAttachment::layout()
-        return baselineElementTop + baselineElementRect.height();
+        if (auto* baselineElementRenderBox = baselineElement->renderBox())
+            return (height() + baselineElementRenderBox->height()) / 2;
+        return height();
```
```suggestion
        // Fallback to the bottom of the attachment if there is no icon.
        return height();
```

squelart committed:
Wide-layout attachments: Fix baseline computation
https://bugs.webkit.org/show_bug.cgi?id=259039
rdar://problem/111991128

Reviewed by Tim Nguyen.

The previous computation used `renderRect`s, which relied on the baseline, modifying the baseline again
and again.
Instead, assume that the image is vertically centred, so it's just a matter of adding half-heights to find
the bottom of that image. (Fallback to the full attachment height alone in case the inner element doesn't
have its height yet.)

* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/rendering/RenderAttachment.cpp:
(WebCore::RenderAttachment::baselinePosition const):

Canonical link: https://commits.webkit.org/265894@main

alanbaradlay commented on Source/WebCore/rendering/RenderAttachment.cpp:
```diff
@@ -97,16 +97,15 @@ void RenderAttachment::layout()
-        if (auto* baselineElementRenderBox = baselineElement->renderBoxModelObject())
-            return baselineElementTop + baselineElementRenderBox->baselinePosition(fontBaseline, firstLine, lineDirectionMode, linePositionMode);
-        return baselineElementTop + baselineElementRect.height();
+        if (auto* baselineElementRenderBox = baselineElement->renderBox()) {
+            // This is the bottom of the image assuming it is vertically centered.
+            return (height() + baselineElementRenderBox->height()) / 2;
```
does wide-attachment support vertical writing mode in general?

squelart replied:
Not at this time, no.
